File: src/mac/dialog.m
// Alerts and file choosers, every one of them asynchronous.
//
// The answer arrives as an event carrying the token the request was made
// with, because a modal loop that blocked here would block the Beans side
// inside a call it could not return from.

#import "internal.h"

// -------------------------------------------------------------------- dialogs

ctd_status ctd_dialog_open(ctd_handle parent, int32_t kind,
                           const char *title, int32_t title_len,
                           const char *body, int32_t body_len,
                           int64_t token) {
    NSWindow *window = nil;
    if (parent) {
        id object = ctd_resolve(parent);
        if (!object) return CTD_ERR_STALE;
        if (![object isKindOfClass:[NSWindow class]]) return CTD_ERR_KIND;
        window = (NSWindow *)object;
        // A sheet needs a window that is actually on screen. Attaching one to
        // a window that was never shown runs no completion handler at all, so
        // the caller waits on a token that will never arrive — a hang, and the
        // worst possible failure for an asynchronous API. A window that cannot
        // host a sheet is treated as no window, which answers immediately.
        if (![window isVisible]) window = nil;
    }
    NSString *heading = ctd_string(title, title_len);
    NSString *detail = ctd_string(body, body_len);

    // The answer is delivered as an event and never returned, because every
    // platform's dialog is asynchronous and a blocking form would have to spin
    // an inner event loop — re-entering the render this call came out of.
    void (^answer)(NSInteger, NSString *) = ^(NSInteger which, NSString *path) {
        ctd_event event;
        memset(&event, 0, sizeof event);
        event.kind = CTD_EV_POST;
        event.token = token;
        event.index = (int64_t)which;
        const char *utf8 = path ? [path UTF8String] : NULL;
        event.text = utf8;
        event.text_len = utf8 ? (int32_t)strlen(utf8) : 0;
        if (g_sink) g_sink(g_sink_context, &event);
    };

    if (kind == CTD_DLG_MESSAGE || kind == CTD_DLG_CONFIRM) {
        NSAlert *alert = [[NSAlert alloc] init];
        [alert setMessageText:heading];
        [alert setInformativeText:detail];
        [alert addButtonWithTitle:@"OK"];
        if (kind == CTD_DLG_CONFIRM) [alert addButtonWithTitle:@"Cancel"];
        if (window) {
            [alert beginSheetModalForWindow:window
                          completionHandler:^(NSModalResponse response) {
                answer(response == NSAlertFirstButtonReturn ? 0 : 1, nil);
            }];
        } else {
            // No surface to attach to — which is the headless case, where
            // running a modal alert would hang. The answer is the default
            // button, reported the same way, so a caller's code path is the
            // same with and without a display.
            answer(0, nil);
        }
        [alert release];
        return CTD_OK;
    }

    if (kind == CTD_DLG_OPEN || kind == CTD_DLG_SAVE) {
        if (!window) {
            // Same reason: a file panel with nothing to attach to would run
            // modally and never return under a test. Report a cancel.
            answer(1, nil);
            return CTD_OK;
        }
        NSSavePanel *panel = kind == CTD_DLG_OPEN
            ? (NSSavePanel *)[NSOpenPanel openPanel]
            : [NSSavePanel savePanel];
        [panel setTitle:heading];
        [panel setMessage:detail];
        [panel beginSheetModalForWindow:window
                      completionHandler:^(NSModalResponse response) {
            if (response != NSModalResponseOK) { answer(1, nil); return; }
            answer(0, [[panel URL] path]);
        }];
        return CTD_OK;
    }
    return CTD_ERR_RANGE;
}

File: src/win32/dialog.c
// Alerts and file choosers, every one of them asynchronous.

#include "internal.h"

typedef struct {
    int64_t token;
    int32_t kind;
    HWND    parent;
    WCHAR  *title;
    WCHAR  *body;
} CtdDialogRequest;

void ctd_run_dialog(void *data) {
    CtdDialogRequest *request = (CtdDialogRequest *)data;
    int64_t index = 0;
    WCHAR path[MAX_PATH];
    path[0] = 0;

    int visible = request->parent && IsWindowVisible(request->parent);
    if (g_role == CTD_ROLE_HEADLESS || !visible) {
        // With nothing on screen to attach to, a dialog answers its default
        // rather than putting a window somewhere nobody asked for. The contract
        // that matters is that **a dialog always answers**: a caller waiting on
        // a token that never arrives is worse than an answer it can see.
        index = (request->kind == CTD_DLG_MESSAGE ||
                 request->kind == CTD_DLG_CONFIRM) ? 0 : 1;
    } else if (request->kind == CTD_DLG_MESSAGE || request->kind == CTD_DLG_CONFIRM) {
        UINT buttons = request->kind == CTD_DLG_CONFIRM ? MB_OKCANCEL : MB_OK;
        int answer = MessageBoxW(request->parent, request->body, request->title,
                                 buttons | MB_ICONINFORMATION);
        index = answer == IDCANCEL ? 1 : 0;
    } else {
        OPENFILENAMEW chooser;
        memset(&chooser, 0, sizeof chooser);
        chooser.lStructSize = sizeof chooser;
        chooser.hwndOwner = request->parent;
        chooser.lpstrFile = path;
        chooser.nMaxFile = MAX_PATH;
        chooser.lpstrTitle = request->title;
        chooser.Flags = OFN_EXPLORER | OFN_NOCHANGEDIR |
                        (request->kind == CTD_DLG_OPEN ? OFN_FILEMUSTEXIST
                                                       : OFN_OVERWRITEPROMPT);
        BOOL chose = request->kind == CTD_DLG_OPEN ? GetOpenFileNameW(&chooser)
                                                   : GetSaveFileNameW(&chooser);
        index = chose ? 0 : 1;
        if (!chose) path[0] = 0;
    }

    if (g_sink) {
        char *text = NULL;
        int32_t text_len = 0;
        if (path[0]) {
            int needed = WideCharToMultiByte(CP_UTF8, 0, path, -1, NULL, 0, NULL, NULL);
            if (needed > 1) {
                text = (char *)malloc((size_t)needed);
                if (text) {
                    WideCharToMultiByte(CP_UTF8, 0, path, -1, text, needed, NULL, NULL);
                    text_len = needed - 1;
                }
            }
        }
        ctd_event event;
        memset(&event, 0, sizeof event);
        event.kind = CTD_EV_POST;
        event.token = request->token;
        event.index = index;
        event.text = text;
        event.text_len = text_len;
        g_sink(g_sink_context, &event);
        free(text);
    }
    free(request->title);
    free(request->body);
    free(request);
}

ctd_status ctd_dialog_open(ctd_handle parent, int32_t kind,
                           const char *title, int32_t title_len,
                           const char *body, int32_t body_len,
                           int64_t token) {
    if (kind < CTD_DLG_MESSAGE || kind > CTD_DLG_SAVE) return CTD_ERR_RANGE;
    HWND owner = NULL;
    if (parent) {
        ctd_status problem;
        owner = ctd_surface_window(parent, &problem);
        if (!owner) return problem;
    }
    CtdDialogRequest *request =
        (CtdDialogRequest *)calloc(1, sizeof(CtdDialogRequest));
    if (!request) return CTD_ERR_PLATFORM;
    request->token = token;
    request->kind = kind;
    request->parent = owner;
    request->title = ctd_wide(title, title_len);
    request->body = ctd_wide(body, body_len);
    if (!request->title || !request->body) {
        free(request->title); free(request->body); free(request);
        return CTD_ERR_PLATFORM;
    }
    // Posted, not run here. `MessageBox` and the file choosers each run a loop
    // of their own, and running one inside this call would re-enter everything
    // above it — including the render the call came out of.
    if (g_role == CTD_ROLE_HEADLESS || !g_limbo) {
        ctd_run_dialog(request);
        return CTD_OK;
    }
    PostMessageW(g_limbo, CTD_WM_DIALOG, 0, (LPARAM)request);
    return CTD_OK;
}
